db/bigtable.go

```diff
@@ -1542,279 +1831,400 @@ func (bigtable *Bigtable) SaveValidatorIncomeDetails(epoch uint64, rewards map[u
 	return nil
 }
 
-func (bigtable *Bigtable) GetEpochIncomeHistoryDescending(startEpoch uint64, endEpoch uint64) (*itypes.ValidatorEpochIncome, error) {
-	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(time.Second*30))
-	defer cancel()
+func (bigtable *Bigtable) MigrateEpochSchemaV1ToV2(epoch uint64) error {
+	funcStart := time.Now()
 
-	ranges := bigtable.getEpochRanges(startEpoch, endEpoch)
-	family := gcp_bigtable.FamilyFilter(STATS_COLUMN_FAMILY)
-	columnFilter := gcp_bigtable.ColumnFilter(SUM_COLUMN)
-	filter := gcp_bigtable.RowFilter(gcp_bigtable.ChainFilters(family, columnFilter))
+	defer func() {
+		logger.Infof("migration of epoch %v completed in %v", epoch, time.Since(funcStart))
+	}()
 
-	res := itypes.ValidatorEpochIncome{}
+	// start := time.Now()
 
-	err := bigtable.tableBeaconchain.ReadRows(ctx, ranges, func(r gcp_bigtable.Row) bool {
-		if len(r[STATS_COLUMN_FAMILY]) == 0 {
-			return false
-		}
-		err := proto.Unmarshal(r[STATS_COLUMN_FAMILY][0].Value, &res)
+	type validatorEpochData struct {
+		ValidatorIndex           uint64
+		Proposals                map[uint64]uint64
+		AttestationTargetSlot    uint64
+		AttestationInclusionSlot uint64
+		SyncParticipation        map[uint64]uint64
+		EffectiveBalance         uint64
+		Balance                  uint64
+		IncomeDetails            *itypes.ValidatorEpochIncome
+	}
+
+	epochData := make(map[uint64]*validatorEpochData)
+	filter := gcp_bigtable.LatestNFilter(1)
+	ctx := context.Background()
+
+	prefixEpochRange := gcp_bigtable.PrefixRange(fmt.Sprintf("%s:e:b:%s", bigtable.chainId, bigtable.reversedPaddedEpoch(epoch)))
+
+	err := bigtable.tableBeaconchain.ReadRows(ctx, prefixEpochRange, func(r gcp_bigtable.Row) bool {
+		// logger.Infof("processing row %v", r.Key())
+
+		keySplit := strings.Split(r.Key(), ":")
+
+		rowKeyEpoch, err := strconv.ParseUint(keySplit[3], 10, 64)
 		if err != nil {
-			logger.Errorf("error decoding income data for row %v: %v", r.Key(), err)
+			logger.Errorf("error parsing epoch from row key %v: %v", r.Key(), err)
 			return false
 		}
+
+		rowKeyEpoch = MAX_EPOCH - rowKeyEpoch
+
+		if epoch != rowKeyEpoch {
+			logger.Errorf("retrieved different epoch than requested, requested: %d, retrieved: %d", epoch, rowKeyEpoch)
```

Comment:
**Suggestion:** Since this function will not be executed periodically, I suggest to add AS MUCH information as possible to every single log entry. "Spam" is not an issue if this is only done once.

Here, for example, why not add the complete row key too?

Also, maybe keep logs like https://github.com/gobitfly/eth2-beaconchain-explorer/blob/ef0180fc0f0250c8c83a06986f59a5affa18e5c3/db/bigtable.go#L1861?